src/bindgen/ir/structure.rs

```diff
@@ -150,6 +150,23 @@ impl Struct {
         }
     }
 
+    /// Attempts to convert this struct to a typedef (only works for transparent structs).
+    pub fn as_typedef(&self) -> Option<Typedef> {
+        if self.is_transparent {
+            // NOTE: A `#[repr(transparent)]` struct with 2+ NZT fields fails to compile, but 0
+            // fields is allowed for some strange reason. Don't emit the typedef in that case.
+            if let Some(field) = self.fields.first() {
```

Comment:
Can you add a test for `#[repr(transparent)] struct Foo;` or so, so that we have test-coverage for the `None` branch?

Reply:
I tried, but it exposed a latent cbindgen bug for test cases that enable struct tagging:
```
Output failed to compile: Output {
  status: ExitStatus(unix_wait_status(256)),
  stdout: "",
  stderr: "
    /path/to/cbindgen/tests/expectations/transparent_tag.compat.c:43:11: 
    error: must use 'struct' tag to refer to type 'TransparentEmptyStructure'
    TransparentEmptyStructure h,
    ^
    struct
    1 error generated.
  "
}
```

If I run the same test against master, cbindgen itself fails:
```
cbindgen failed: Some("/path/to/cbindgen/tests/expectations/transparent.compat.c") with error: 
thread 'main panicked at src/bindgen/language_backend/clike.rs:548:34:
index out of bounds: the len is 0 but the index is 0
```
My newly added check avoided the index out of bounds panic, but I don't know where the tagging issue would lurk?

Reply:
Also filed https://github.com/rust-lang/rust/issues/129029, because it seems like the rust compiler shouldn't allow an empty transparent struct in the first place.

Reply:
Found it: [structure.rs:274](https://github.com/mozilla/cbindgen/blob/58c6156b0d91e82abb03c26187b8d18fa4345ce0/src/bindgen/ir/structure.rs#L274):
```diff
- if self.is_transparent {
+ if self.is_transparent && self.fields.len() == 1 {
```

Reply:
The above fix is surgical, but I'm not sure it's ideal? Should we forbid `is_transparent` for an empty struct, rather than tolerate it like this code currently does? I worry the checks are proliferating...

Reply:
I went ahead and moved the check into the constructor, so that transparent implies single field everywhere else.

Reply:
Hmm. It's apparently intentional behavior to allow an empty transparent struct: https://github.com/rust-lang/rust/issues/77841#issuecomment-716575747. But I don't know what the underlying should be in that case, if we were to emit a typedef. Should we continue suppressing the typedef for empty transparent structs? Or is there some canonical type we should use as underlying for the typedef?

Reply:
Update: All zero-sized structs now trigger a warning about undefined behavior, and an empty struct definition is emitted regardless of whether the user asked for repr(C) or repr(transparent).